core/orchestrator.go

```diff
@@ -536,8 +536,38 @@ func (rtm *RemoteTranscoderManager) RegisteredTranscodersInfo() []net.RemoteTran
 }
 
 func (rtm *RemoteTranscoderManager) Manage(stream net.Transcoder_RegisterTranscoderServer, capacity int) {
-
+	from := common.GetConnectionAddr(stream.Context())
 	transcoder := NewRemoteTranscoder(rtm, stream, capacity)
+	go func() {
+		for {
+			// Monitor connection state
+			_, err := stream.Recv()
+			if err != nil {
+				cur := atomic.LoadInt64(&transcoder.transcodingSegments)
+				glog.V(common.DEBUG).Infof("Got error from remote transcoder=%s transcoding segments=%d stream err=%v", from, cur, err)
+				if err == io.EOF {
+					glog.Infof("Performing gradual shutdown of transcoder=%s", from)
+					// Transcoder wants to do gradual shutdown
+					// remove it from map so no tasks assigned
+					// and wait till all current transcoding tasks will finish
+					rtm.RTmutex.Lock()
+					delete(rtm.liveTranscoders, transcoder.stream)
+					rtm.RTmutex.Unlock()
+					for {
+						if atomic.LoadInt64(&transcoder.transcodingSegments) == 0 {
+							break
+						}
+						time.Sleep(100 * time.Millisecond)
```

Comment:
This feels really fragile. If we are to handle this case at all, it seems better to do it on the client side: upon a graceful disconnect, reject new incoming segments, then close the stream once all segments are complete.

Reply:
Actually maybe we can close the stream on the client by cancelling the context (rather than having an explicit reject), then holding off on the shutdown until all segments are complete. T can continue segment processing as normal. This should work since the request to transcode the segment is separate from the segment notification request.

My guess is that a `sync.WaitGroup` for each transcode should do the trick, and doing a `wg.Wait()` after the `Recv` for-loop  within `runTranscoder`.